### PR TITLE
GCC-11: Set DWARF to version 4 and ignore array bound warnings for generated syscalls functions

### PR DESCRIPTION
**Pull request by dcpleung**
* GCC 11 defaults to output DWARF version 5 which pyelftools cannot currently parse. So keep output at version 4 for the time being.
* Compilers and static code analyzers do not understand Zephyr's syscall mechanism so they always complain about out of bound array access inside the generated syscall header functions. So add a flag for GCC to ignore this type of warning for these functions.

Related to #35707

### CODE REVIEW

**Comment by dcpleung** [comment]
@tejlmand I am not sure about where to put `-gdwarf-4`. Please let me know if it is incorrect so I can move it.

**Review by galak — APPROVED** [review]

File: cmake/compiler/gcc/compiler_flags.cmake [hunk 1]
```diff
@@ -166,6 +166,7 @@ set_compiler_property(PROPERTY freestanding -ffreestanding)
 
 # Flag to enable debugging
 set_compiler_property(PROPERTY debug -g)
+set_compiler_property(PROPERTY debug -gdwarf-4)
```
Comment (tejlmand) [inline cmake/compiler/gcc/compiler_flags.cmake:169]: Although dwarf-4 standard is from 2010 should we use a `check_set_compiler_property()` to ensure that we are not breaking things for proprietary compilers that are based on an old gcc ?

Comment (dcpleung) [reply cmake/compiler/gcc/compiler_flags.cmake]: Updated to use `check_set_compiler_property()`.

According to code, pyelftools supports version 2 to 4.

**Review by tejlmand — APPROVED** [review]

File: cmake/compiler/gcc/compiler_flags.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/compiler/gcc/compiler_flags.cmake]: Thanks for checking.

**Review by enjiamai — APPROVED** [review]

**Pull request merged**